File: Documentospdf/informeMedico.js
// informeMedico.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { getProfessionalData } from "./professionalResolver.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export function generarInformeMedico(doc, datos) {
  const {
    nombre,
    edad,
    rut,
    motivoConsulta,
    antecedentes,
    examenFisico,
    estudios,
    impresionDiagnostica,
    plan,
    professional // ← viene del login
  } = datos || {};

  const medico = getProfessionalData(professional);

  if (!medico) {
    throw new Error("Profesional no encontrado");
  }

  /* ================= ENCABEZADO ================= */
  try {
    const logoPath = path.join(__dirname, "assets", "ica.jpg");
    if (fs.existsSync(logoPath)) {
      doc.image(logoPath, 50, 40, { width: 120 });
    }
  } catch {}

  doc.moveDown(1.5);
  doc.font("Helvetica-Bold").fontSize(18)
    .text("INSTITUTO DE CIRUGÍA ARTICULAR", 180, 50);

  doc.moveDown(1.5);
  doc.fontSize(16)
    .text("INFORME MÉDICO", 180, undefined, {
      underline: true,
    });

  doc.moveDown(4);
  doc.x = doc.page.margins.left;

  /* ================= DATOS PACIENTE ================= */
  doc.font("Helvetica").fontSize(13);
  doc.text(`Nombre: ${nombre ?? ""}`);
  doc.moveDown(0.8);
  doc.text(`Edad: ${edad ?? ""}`);
  doc.moveDown(0.5);
  doc.text(`RUT: ${rut ?? ""}`);
  doc.moveDown(1.5);

  /* ================= CONTENIDO CLÍNICO ================= */

  function seccion(titulo, contenido) {
    if (contenido && contenido.trim()) {
      doc.moveDown(1);
      doc.font("Helvetica-Bold").fontSize(14).text(titulo);
      doc.moveDown(0.5);
      doc.font("Helvetica").fontSize(12).text(contenido.trim(), {
        align: "justify"
      });
    }
  }

  seccion("Motivo de Consulta", motivoConsulta);
  seccion("Antecedentes Relevantes", antecedentes);
  seccion("Examen Físico", examenFisico);
  seccion("Estudios Complementarios", estudios);
  seccion("Impresión Diagnóstica", impresionDiagnostica);
  seccion("Plan y Conducta", plan);

  /* ================= FIRMA DINÁMICA ================= */
  const pageW = doc.page.width;
  const pageH = doc.page.height;
  const marginL = doc.page.margins.left || 50;
  const marginR = doc.page.margins.right || 50;
  const baseY = pageH - 170;

  doc.font("Helvetica").fontSize(12);
  doc.text("_________________________", marginL, baseY, {
    align: "center",
    width: pageW - marginL - marginR,
  });
  doc.text("Firma y Timbre Médico", marginL, baseY + 18, {
    align: "center",
    width: pageW - marginL - marginR,
  });

  // Firma dinámica
  try {
    const firmaPath = path.join(__dirname, "assets", medico.firma);
    if (fs.existsSync(firmaPath)) {
      const firmaW = 250;
      const firmaX = (pageW - firmaW) / 2;
      const firmaY = baseY - 45;
      doc.image(firmaPath, firmaX, firmaY, { width: firmaW });
    }
  } catch {}

  // Timbre dinámico
  try {
    const timbrePath = path.join(__dirname, "assets", medico.timbre);
    if (fs.existsSync(timbrePath)) {
      const firmaW = 250;
      const firmaX = (pageW - firmaW) / 2;
      const timbreW = 110;
      const timbreX = firmaX + firmaW;
      const timbreY = baseY - 65;

      doc.save();
      doc.rotate(20, {
        origin: [timbreX + timbreW / 2, timbreY + timbreW / 2],
      });
      doc.image(timbrePath, timbreX, timbreY, { width: timbreW });
      doc.restore();
    }
  } catch {}

  /* ================= PIE DINÁMICO ================= */
  doc.font("Helvetica").fontSize(12);
  doc.text(medico.nombre, marginL, baseY + 52, {
    align: "center",
    width: pageW - marginL - marginR,
  });
  doc.text(`RUT: ${medico.rut}`, {
    align: "center",
    width: pageW - marginL - marginR,
  });
  doc.text(medico.especialidad, {
    align: "center",
    width: pageW - marginL - marginR,
  });
  doc.text("INSTITUTO DE CIRUGÍA ARTICULAR", {
    align: "center",
    width: pageW - marginL - marginR,
  });
}

File: Documentospdf/informeMedico.py
import os
from datetime import datetime, date
from reportlab.lib.pagesizes import A4
from reportlab.pdfgen import canvas
from reportlab.lib.utils import ImageReader
from Documentospdf.professionalResolver import getProfessionalData


def calcular_edad(fecha_nacimiento_str):
    try:
        nacimiento = datetime.strptime(fecha_nacimiento_str, "%Y-%m-%d").date()
        hoy = date.today()
        return hoy.year - nacimiento.year - (
            (hoy.month, hoy.day) < (nacimiento.month, nacimiento.day)
        )
    except:
        return ""


def wrap_text(text, font_name, font_size, max_width, canvas_obj):
    """Divide texto en líneas que no excedan max_width."""
    canvas_obj.setFont(font_name, font_size)
    words  = text.split()
    lines  = []
    current = ""

    for word in words:
        test = (current + " " + word).strip()
        if canvas_obj.stringWidth(test, font_name, font_size) <= max_width:
            current = test
        else:
            if current:
                lines.append(current)
            current = word

    if current:
        lines.append(current)

    return lines


def draw_wrapped_text(c, text, x, y, font_name, font_size, max_width, leading, min_y):
    """
    Dibuja texto con wrap. Crea nueva página si se acaba el espacio.
    Devuelve la posición Y final.
    """
    paragraphs = text.split("\n")
    c.setFont(font_name, font_size)

    for para in paragraphs:
        if not para.strip():
            y -= leading * 0.5
            continue

        lines = wrap_text(para, font_name, font_size, max_width, c)

        for line in lines:
            if y < min_y:
                c.showPage()
                y = A4[1] - 60
                c.setFont(font_name, font_size)

            c.drawString(x, y, line)
            y -= leading

        y -= leading * 0.3  # espacio entre párrafos

    return y


def generar_informe_pdf(buffer, datos):

    # =========================
    # PACIENTE
    # =========================
    nombre           = datos.get("nombre", "")
    apellido_paterno = datos.get("apellido_paterno", "")
    apellido_materno = datos.get("apellido_materno", "")
    nombre_completo  = f"{nombre} {apellido_paterno} {apellido_materno}".strip()

    rut        = datos.get("rut", "")
    diagnostico = datos.get("diagnostico", "")
    texto_rp   = datos.get("indicaciones", "")
    professional = datos.get("professional")

    edad = datos.get("edad")
    if not edad and datos.get("fecha_nacimiento"):
        edad = calcular_edad(datos.get("fecha_nacimiento"))
    if not edad:
        edad = ""

    # =========================
    # PROFESIONAL
    # =========================
    medico = getProfessionalData(professional)
    if not medico:
        raise Exception("Profesional no encontrado")

    # =========================
    # CANVAS
    # =========================
    c = canvas.Canvas(buffer, pagesize=A4)
    width, height = A4

    current_dir = os.path.dirname(os.path.abspath(__file__))
    assets_dir  = os.path.abspath(os.path.join(current_dir, "..", "assets"))

    MARGIN_LEFT  = 60
    MARGIN_RIGHT = 60
    TEXT_WIDTH   = width - MARGIN_LEFT - MARGIN_RIGHT  # ~475pt
    MIN_Y        = 160  # espacio para firma

    # =========================
    # LOGO
    # =========================
    logo_path = os.path.join(assets_dir, "ica.jpg")
    if os.path.exists(logo_path):
        c.drawImage(
            ImageReader(logo_path),
            MARGIN_LEFT, height - 155,
            width=110, height=110,
            preserveAspectRatio=True, mask="auto"
        )

    # =========================
    # ENCABEZADO
    # =========================
    c.setFont("Helvetica-Bold", 14)
    c.drawString(200, height - 55, "INSTITUTO DE CIRUGÍA ARTICULAR")

    c.setFont("Helvetica-Bold", 16)
    c.drawString(200, height - 78, "INFORME MEDICO")

    y = height - 170

    # =========================
    # DATOS PACIENTE
    # =========================
    c.setFont("Helvetica", 12)
    c.drawString(MARGIN_LEFT, y, f"Nombre: {nombre_completo}")
    y -= 24

    c.drawString(MARGIN_LEFT, y, f"Edad: {edad}")
    y -= 24

    c.drawString(MARGIN_LEFT, y, f"RUT: {rut}")
    y -= 24

    # Diagnóstico con wrap
    diag_lines = wrap_text(f"Diagnóstico: {diagnostico}", "Helvetica", 12, TEXT_WIDTH, c)
    for line in diag_lines:
        c.drawString(MARGIN_LEFT, y, line)
        y -= 20
    y -= 20

    # =========================
    # Rp.
    # =========================
    c.setFont("Helvetica-Bold", 18)
    c.drawString(MARGIN_LEFT, y, "Rp.")
    y -= 40

    # =========================
    # TEXTO CLÍNICO CON WRAP
    # =========================
    if texto_rp:
        y = draw_wrapped_text(
            c, texto_rp,
            x=MARGIN_LEFT + 15,
            y=y,
            font_name="Helvetica",
            font_size=12,
            max_width=TEXT_WIDTH - 15,
            leading=18,
            min_y=MIN_Y
        )
    else:
        c.setFont("Helvetica", 12)
        c.drawString(MARGIN_LEFT + 15, y, "____________________________")

    # =========================
    # FIRMA Y TIMBRE
    # =========================
    baseY = 110

    firma_path = os.path.join(assets_dir, medico.get("firma", ""))
    if os.path.exists(firma_path):
        c.drawImage(
            ImageReader(firma_path),
            width / 2 - 110, baseY + 30,
            width=220, height=85,
            preserveAspectRatio=True, mask="auto"
        )

    timbre_path = os.path.join(assets_dir, medico.get("timbre", ""))
    if os.path.exists(timbre_path):
        c.saveState()
        timbre_x = width / 2 + 95
        timbre_y = baseY + 35
        tw = th = 95
        c.translate(timbre_x + tw / 2, timbre_y + th / 2)
        c.rotate(-20)
        c.drawImage(
            ImageReader(timbre_path),
            -tw / 2, -th / 2,
            width=tw, height=th,
            preserveAspectRatio=True, mask="auto"
        )
        c.restoreState()

    # =========================
    # LÍNEA Y NOMBRE PROFESIONAL
    # =========================
    c.setFont("Helvetica", 11)
    c.drawCentredString(width / 2, baseY,      "_____________________________________")
    c.drawCentredString(width / 2, baseY - 16, medico.get("name", ""))
    c.drawCentredString(width / 2, baseY - 30, medico.get("specialty", ""))
    c.drawCentredString(width / 2, baseY - 44, "INSTITUTO DE CIRUGÍA ARTICULAR")

    c.showPage()
    c.save()
